### PR TITLE
Use Fragments for Figure Captions, removing Text component

### LINKED ISSUE #801: Use fragmented data to render rich text for image captions
**Is your feature request related to a problem? Please describe.**
With work being merged to render fragment data with #672, the [image](https://github.com/BBC-News/simorgh/blob/latest/src/app/containers/Image/index.jsx) components will render plain text rather than rich text. This is because the [`model.text` property in Fragmented data]() is just plain text.

**Describe the solution you'd like**
We should change the Heading component to use the [Fragment](https://github.com/BBC-News/simorgh/blob/latest/src/app/containers/Fragment/index.jsx) component to render bold, italics, and all other rich text. There are 2 options for using the Fragment component, and there needs to be a consideration for what option to use. These options are:
- Use the Fragment component 'directly' with the [Blocks](https://github.com/BBC-News/simorgh/blob/latest/src/app/containers/Blocks/index.jsx) component (`Caption -> Blocks -> Fragment`) --- This means would need to extract `Fragment` out of each `Text -> Paragraph` block, which might be complex to maintain
- Use the Text component. It currently renders `Text -> Blocks -> Paragraph -> Blocks -> Fragment`, and would need a refactor to be able to just return the string of text that can be nested inside the Caption component, i.e. `Caption -> Text -> Blocks -> Raw Text String(?) -> Blocks -> Fragment` --- This would add cognitive complexity to the rendering structure, which would include CPU overhead

As part of this work, we should add a Figure Caption story which contains a link, defined in fragment style (Following on from #1017).

**Describe alternatives you've considered**
See 2 potential solutions in descriptions.

**Testing notes**
Will Cypress tests be required or are unit tests sufficient?

- [x] Intially labelled with ["Refinement needed"](https://github.com/BBC-News/simorgh/labels/Refinement%20Needed)

### LINKED ISSUE #1016: Remove Markdown-to-JSX
**Is your feature request related to a problem? Please describe.**
When solving text formatting, we initially went down the markdown route before realising it would not scale for i18n.

Optimo and Ares now return fragment blocks, i.e. text is associated with an `attributes` array which might contain `italic`, `bold`, etc. It no longer returns markdown such as `_italic_` or `*bold*`.

For legacy reasons, we still have markdown rendering in our Text component, which means we go from attributes -> markdown -> JSX. We should just go from attributes -> JSX directly.

**Describe the solution you'd like**

1. `npm uninstall markdown-to-jsx`
2. Update `components/Text` to not use Markdown anymore.
3. Update `containers/Text` to pull in `containers/InlineLink/index.new.jsx` (as opposed to `index.jsx`)
4. Update Text README.md.
5. Remove [reference to markdown in Caption comment](https://github.com/BBC-News/simorgh/blob/6de3063b80084f7017cea0554c86d8d01339bb9c/src/app/components/Figure/Caption/index.jsx#L40).

NB: we'll also want to make sure we have no markdown in any of our `data/`.

**Describe alternatives you've considered**
N/A

**Testing notes**
[Tester to complete]

Dev insight: Regression tests only.

- [x] Initially labelled with ["Refinement needed"](https://github.com/BBC-News/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by bcmn**
Resolves #801 & resolves #1016. Merge blocked on #1030.

- Removes usages of Text component from the HorizontalRule story & the Figure Caption container.
- Removes Text component & `markdown-to-jsx`. 🎉 
- Updates Image container to use Blocks to use fragments for rich text Figure Captions.
- Re-adds stories & tests for figure captions containing links, using new non-Markdown inline link definitions.
- In adding those stories & tests, updates the `models/blocks.js` to reflect that unstyled text should still contain a single fragment. This causes a number of seemingly unrelated snapshot changes.

For testing/viewing changes, it's worth noting that the _local_ scenarios 8 & 20 contain image captions with urlLinks. The scenarios on TEST do not yet, due to an issue with Optimo.

- [ ] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

**Comment by bcmn** [comment]
[Phil's raised a interesting point here](https://github.com/BBC-News/simorgh/issues/1028#issuecomment-443680508) about whether or not we should implement the block-based captions given we know the UX requirements are subject to change. There's one commit in it, so if we can come to a consensus the rest of this PR should still be able to move reasonably quickly, either way. ¯\_(ツ)_/¯

File: package.json [hunk 1]
```diff
@@ -71,7 +71,6 @@
     "gnu-terry-pratchett": "0.0.2",
     "helmet": "^3.14.0",
     "isomorphic-fetch": "^2.2.1",
-    "markdown-to-jsx": "^6.8.3",
```
Comment (ChrisBAshton) [inline package.json:76]: 🎉

File: src/app/containers/Image/index.jsx [hunk 2]
```diff
@@ -2,16 +2,26 @@ import React from 'react';
 import { filterForBlockType } from '../../helpers/blockHandlers';
 import { imageModelPropTypes } from '../../models/propTypes/image';
 import Figure from '../Figure';
+import Blocks from '../Blocks';
+import fragment from '../Fragment';
```
Comment (ChrisBAshton) [inline src/app/containers/Image/index.jsx]: Shouldn't this be `Fragment` uppercased like the others?

File: src/app/containers/Image/index.jsx [hunk 3]
```diff
@@ -2,16 +2,26 @@ import React from 'react';
...
+import InlineLink from '../InlineLink';
 
 const DEFAULT_IMAGE_RES = 640;
 
 const getText = ({ model }) => model.blocks[0].model.blocks[0].model.text;
 
+const componentsToRender = { fragment, urlLink: InlineLink };
```
Comment (ChrisBAshton) [inline src/app/containers/Image/index.jsx]: ...oh. I see why. But still: `fragment: Fragment`

File: src/app/containers/Image/index.jsx [hunk 4]
```diff
@@ -2,16 +2,26 @@ import React from 'react';
...
 const getCaption = block => {
   if (!block) {
     return null;
   }
-  return getText(block);
+
+  return (
+    <Blocks
+      blocks={block.model.blocks[0].model.blocks[0].model.blocks}
```
Comment (ChrisBAshton) [inline src/app/containers/Image/index.jsx]: As discussed in Slack - there _must_ be a way to refactor this 🤔but that's for a future PR.

**Review by ChrisBAshton — DISMISSED** [review]
Looks good to me. RE the comment around whether or not we should be implementing support for inline links inside captions - let's discuss as a team tomorrow.

**Comment by pjlee11** [comment]
⚠️ 

Going to raise [this](https://github.com/BBC-News/simorgh/issues/1028#issuecomment-443767799) again, I think we should be very aware we will be integrated known styling bugs if we support using inline links in the caption before implementing the new UX designs.

**Comment by pjlee11** [comment]
Bug examples with current UX designs when using an inline link in a caption
 
image

image

image

image

**Comment by ChrisBAshton** [comment]
Thanks for the info, Phil - let's discuss as a team tomorrow.

File: src/app/containers/Image/index.jsx (hunk 3 above)
Comment (bcmn) [reply src/app/containers/Image/index.jsx]: Yeah, I think I was just following similar usages. Fixed.

**Comment by jimjohnsonrollings** [comment]
My view is that we merge this as-is and create a new issue to fix the UX once the designs are updated.

**Comment by ChrisBAshton** [comment]
Agreed in meeting we'll merge as-is and raise a new issue to fix the rendering issues outlined here.

**Review by ChrisBAshton — DISMISSED** [review]

File: src/app/containers/Image/index.jsx [hunk 5]
```diff
@@ -2,16 +2,26 @@ import React from 'react';
...
+const captionComponentsToRender = { fragment: Fragment, urlLink: InlineLink };
 const getCaption = block => {
   if (!block) {
     return null;
   }
-  return getText(block);
+
+  return (
+    <Blocks
+      blocks={block.model.blocks[0].model.blocks[0].model.blocks}
```
Comment (dr3) [inline src/app/containers/Image/index.jsx]: Could this be moved inside caption container? Doesn't seem very image specific and will have to be moved anyway when we have video

File: src/app/containers/Figure/index.jsx [hunk 6]
```diff
@@ -27,7 +27,7 @@ FigureContainer.propTypes = {
   src: string.isRequired,
   ratio: number.isRequired,
   copyright: string,
-  caption: string,
+  caption: node,
```
Comment (dr3) [inline src/app/containers/Figure/index.jsx]: Regardless of which way we do this, I think we should distinquish `caption` and `captionContents` as this `node` isnt the caption, its whats inside it. Not very clear imo

**Review by dr3 — CHANGES_REQUESTED** [review]

**Comment by bcmn** [comment]
Blocking this on #1044, as achieving Drew's requested changes would involve updating files that will be changed by that PR.

**Comment by bcmn** [comment]
Just capturing what happened in the last few commits: ☝🏻

I moved the logic for rendering caption text as `<Blocks>` to the caption container, following Drew's request. This means the Figure & Caption containers now take block objects, not React nodes.

In testing this change, I realised `models/blocks.js` is a little bit out of date, & has very little awareness of fragments as something we need to contend with in our data feed. As such, I had to update it to include a single fragment for its' unstyled text block. Unfortunately, this means several other snapshots changed, making this PR slightly noisier.

I then re-added the "caption containing inline link" stories & tests, following #1017 & a nudge from James. 😛 This could really benefit from a refactor of `models/blocks.js`, as it now means there is a block definition in `containers/figure/fixtureData.js`, but doing such a refactor is out of scope for this PR. I'll raise a ticket.

File: src/app/containers/Image/index.jsx (hunk 5 above)
Comment (bcmn) [reply src/app/containers/Image/index.jsx]: Should be addressed in following commits.

File: src/app/containers/Figure/index.jsx (hunk 6 above)
Comment (bcmn) [reply src/app/containers/Figure/index.jsx]: Should be addressed in following commits.

**Comment by ChrisBAshton** [comment]
Ticket raised for ^^ is #1052

**Comment by bcmn** [comment]
No longer blocked on #1044, which has been merged.

**Review by ChrisBAshton — DISMISSED** [review]

File: src/app/containers/Caption/index.jsx [hunk 7]
```diff
@@ -1,33 +1,35 @@
...
 import VisuallyHiddenText from '../../components/VisuallyHiddenText';
+import Blocks from '../Blocks';
+import Fragment from '../Fragment';
+import InlineLink from '../InlineLink';
 
-const CaptionWithOffscreenText = ({ children }) => (
+const componentsToRender = { fragment: Fragment, urlLink: InlineLink };
+
+const renderText = block => {
+  const textBlocks = block.model.blocks[0].model.blocks[0].model.blocks;
```
Comment (dr3) [inline src/app/containers/Caption/index.jsx:13]: Do we need to defend against this? feels like a place "could not find X of undefined" could come up

**Review by dr3 — DISMISSED** [review]

File: src/app/containers/Caption/index.jsx (hunk 7 above)
Comment (bcmn) [reply src/app/containers/Caption/index.jsx:13]: Yeah -- as discussed in person, [I reckon this work should be done alongside the refactor Chris suggested above](https://github.com/BBC-News/simorgh/pull/1033#discussion_r238358370), as both aren't exclusive to this PR, but occur in places throughout our containers.

File: src/app/components/Figure/Caption/index.stories.jsx [hunk 8]
```diff
@@ -1,7 +1,16 @@
...
+import InlineLink from '../../InlineLink';
 
-storiesOf('Caption', module).add('default', () => (
-  <Caption>This is a caption.</Caption>
-));
+storiesOf('Caption', module)
+  .add('default', () => <Caption>This is a caption.</Caption>)
+  .add('containing an inline link', () => (
+    <Caption>
+      This is a caption{' '}
```
Comment (pjlee11) [inline src/app/components/Figure/Caption/index.stories.jsx]: What's `{' '}` achieving here?

Comment (ChrisBAshton) [reply src/app/components/Figure/Caption/index.stories.jsx]: Good question!

File: src/app/containers/Caption/index.jsx [hunk 9]
```diff
@@ -1,33 +1,35 @@
...
-  children: node.isRequired,
-};
-
-const CaptionContainer = ({ captionValue }) => (
-  <Text text={captionValue} paragraphOverride={CaptionWithOffscreenText} />
-);
-
 CaptionContainer.propTypes = {
-  captionValue: string.isRequired,
+  block: objectOf(any).isRequired,
```
Comment (pjlee11) [inline src/app/containers/Caption/index.jsx:32]:
```suggestion
  block: node.isRequired,
```

File: src/app/containers/Caption/index.jsx [hunk 10]
```diff
@@ -1,33 +1,35 @@
 import React from 'react';
-import { node, string } from 'prop-types';
+import { objectOf, any } from 'prop-types';
```
Comment (pjlee11) [inline src/app/containers/Caption/index.jsx:2]:
```suggestion
import { node } from 'prop-types';
```

File: src/app/components/Figure/Caption/index.stories.jsx (hunk 8 above)
Comment (sareh) [reply src/app/components/Figure/Caption/index.stories.jsx]: This is what prettier/eslint config does to keep a space after the work `caption` and before the `<InlineLink>` element. You can work around it by saving the text in a variable. e.g. `const captionText = 'This is a caption '` & use it in the element.

File: src/app/containers/Caption/index.jsx (hunk 9 above)
Comment (bcmn) [reply src/app/containers/Caption/index.jsx:32]: We talked about both of these -- it's now passing the JavaScript object for the caption block, not passing the React node for it.

File: src/app/containers/Caption/index.jsx (hunk 10 above)
Comment (bcmn) [reply src/app/containers/Caption/index.jsx:2]: See above.

File: src/app/components/Figure/Caption/index.stories.jsx (hunk 8 above)
Comment (bcmn) [reply src/app/components/Figure/Caption/index.stories.jsx]:
```suggestion
      {'This is a caption '}
```


screenshot 2018-12-07 at 08 38 25
This seems to have the same result, so I'll do that.

**Review by sareh — DISMISSED** [review]

**Review by pjlee11 — DISMISSED** [review]

**Review by ChrisBAshton — APPROVED** [review]

**Review by dr3 — APPROVED** [review]

**Comment by jamesbrumpton** [comment]
LGTM

**Pull request merged**